File: src/main/java/de/qabel/core/config/Accounts.java
package de.qabel.core.config;

import java.util.*;

/**
 * https://github.com/Qabel/qabel-doc/wiki/Qabel-Client-Configuration#accounts
 */
public class Accounts {

    private final Map<String, Account> accounts = new HashMap<>();

    /**
     * Returns unmodifiable set of contained accounts
     *
     * @return Set<Account>
     */
    public Set<Account> getAccounts() {
        return Collections.unmodifiableSet(new HashSet<>(this.accounts.values()));
    }

    /**
     * Put an account
     *
     * @param account Account to put
     * @return True if newly added, false if updated
     */
    public boolean put(Account account) {
        if (this.accounts.put(account.getPersistenceID(), account) == null) {
            return true;
        }
        return false;
    }

    /**
     * Removes account from list of accounts
     *
     * @param account Account to be removed.
     * @return true if account was contained in list, false if not
     */
    public boolean remove(Account account) {
        if (this.accounts.remove(account.getPersistenceID()) == null) {
            return false;
        }
        return true;
    }

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + ((accounts == null) ? 0 : accounts.hashCode());
        return result;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null) {
            return false;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        Accounts other = (Accounts) obj;
        if (accounts == null) {
            if (other.accounts != null) {
                return false;
            }
        } else if (!accounts.equals(other.accounts)) {
            return false;
        }
        return true;
    }
}

File: core/src/main/java/de/qabel/core/config/Accounts.kt
package de.qabel.core.config

import java.util.*

/**
 * https://github.com/Qabel/qabel-doc/wiki/Qabel-Client-Configuration#internalAccounts
 */
class Accounts {

    private val internalAccounts = HashMap<String, Account>()

    /**
     * Returns unmodifiable set of contained internalAccounts

     * @return Set
     */
    fun getAccounts(): Set<Account> {
        return Collections.unmodifiableSet(HashSet(internalAccounts.values))
    }

    /**
     * Put an account

     * @param account Account to put
     * *
     * @return True if newly added, false if updated
     */
    fun put(account: Account): Boolean {
        return internalAccounts.put(account.persistenceID, account) == null
    }

    /**
     * Removes account from list of internalAccounts

     * @param account Account to be removed.
     * *
     * @return true if account was contained in list, false if not
     */
    fun remove(account: Account): Boolean {
        return internalAccounts.remove(account.persistenceID) != null
    }

    override fun hashCode(): Int {
        val prime = 31
        var result = 1
        result = prime * result + if (internalAccounts == null) 0 else internalAccounts.hashCode()
        return result
    }

    override fun equals(obj: Any?): Boolean {
        if (this === obj) {
            return true
        }
        if (obj == null) {
            return false
        }
        if (javaClass != obj.javaClass) {
            return false
        }
        val other = obj as Accounts?
        if (internalAccounts == null) {
            if (other!!.internalAccounts != null) {
                return false
            }
        } else if (internalAccounts != other!!.internalAccounts) {
            return false
        }
        return true
    }
}
